src/core/math_func.hpp

```diff
@@ -307,60 +323,93 @@ int DivideApprox(int a, int b);
- * @param a Numerator
- * @param b Denominator
- * @return Quotient, rounded up
+ * @param a Numerator.
+ * @param b Denominator.
+ * @return Quotient, rounded up.
  */
-static inline uint CeilDiv(uint a, uint b)
+static inline uint CeilDiv(const uint a, const uint b)
 {
 	return (a + b - 1) / b;
 }
 
 /**
  * Computes ceil(a / b) * b for non-negative a and b.
- * @param a Numerator
- * @param b Denominator
+ * @param a Numerator.
+ * @param b Denominator.
  * @return a rounded up to the nearest multiple of b.
  */
-static inline uint Ceil(uint a, uint b)
+static inline uint Ceil(const uint a, const uint b)
 {
 	return CeilDiv(a, b) * b;
 }
 
 /**
- * Computes round(a / b) for signed a and unsigned b.
- * @param a Numerator
- * @param b Denominator
- * @return Quotient, rounded to nearest
+ * Computes round(a / b) for b > 0.
+ * @param a Numerator.
+ * @param b Denominator.
+ * @return Quotient, rounded to nearest.
  */
-static inline int RoundDivSU(int a, uint b)
+static inline int RoundDivSU(const int a, const uint b)
 {
+	if (unlikely(b >= std::numeric_limits<int>::max())) return 0;
+
+	const int b_ = static_cast<int>(b);
 	if (a > 0) {
 		/* 0.5 is rounded to 1 */
-		return (a + static_cast<int>(b) / 2) / static_cast<int>(b);
+		return (a + b_ / 2) / b_;
 	} else {
 		/* -0.5 is rounded to 0 */
-		return (a - (static_cast<int>(b) - 1) / 2) / static_cast<int>(b);
+		return (a - (b_ - 1) / 2) / b_;
 	}
 }
 
 /**
- * Computes (a / b) rounded away from zero.
- * @param a Numerator
- * @param b Denominator
- * @return Quotient, rounded away from zero
+ * Computes (a / b) rounded away from zero for b > 0.
+ * @param a Numerator.
+ * @param b Denominator.
+ * @return Quotient, rounded away from zero.
  */
-static inline int DivAwayFromZero(int a, uint b)
+static inline int DivAwayFromZero(const int a, const uint b)
 {
-	const int _b = static_cast<int>(b);
+	if (unlikely(b >= std::numeric_limits<int>::max())) return signum(a);
+
+	const int b_ = static_cast<int>(b);
 	if (a > 0) {
-		return (a + _b - 1) / _b;
+		return (a + b_ - 1) / b_;
 	} else {
 		/* Note: Behaviour of negative numerator division is truncation toward zero. */
-		return (a - _b + 1) / _b;
+		return (a - b_ + 1) / b_;
 	}
 }
 
+/**
+ * Computes (a / b) rounded towards negative infinity for b > 0.
+ * @param a Numerator.
+ * @param b Denominator.
+ * @return Quotient, rounded towards negative infinity.
+ */
+static inline int DivTowardsNegativeInf(const int a, const uint b)
+{
+	if (unlikely(b >= std::numeric_limits<int>::max())) return min(0, signum(a));
+
+	const int b_ = static_cast<int>(b);
+	return (a / b_) - (a % b_ < 0 ? 1 : 0);
+}
+
+/**
+ * Computes (a / b) rounded towards positive infinity for b > 0.
+ * @param a Numerator.
+ * @param b Denominator.
+ * @return Quotient, rounded towards positive infinity.
+ */
+static inline int DivTowardsPositiveInf(const int a, const uint b)
+{
+	if (unlikely(b >= std::numeric_limits<int>::max())) return max(0, signum(a));
```

Comment:
warning C4018: '>=': signed/unsigned mismatch (MSVC warning)

Reply:
What is the compiler actually complaining about here? Should `std::numeric_limits<int>::max()` be typecast to `uint`?